### DISCUSSION TITLE
How to modify Create 3 Docker image to use a different default RMW

### DISCUSSION THREAD

**Discussion by framlin** [iRobotEducation/create3_docs#276] [Questions] [answered]
Maybe this is a docker - question, but I do not know anything about docker and I need it for my create3, so I ask it here.
I have to use galactic on a Ubuntu 22.04, so I use the docker-image, provided by iRobot for that.
It works perfect, if I use CycloneDDS and I can switch manually to Fast-DDS, too.
But what I would like to have, is to start automatically with Fast-DDS.
Does anybody know, how I could achieve this?

**Comment by alsora** [comment, marked as answer]
Yes, this is really a Docker-related question and a complete answer would require to have some understanding of how Docker works.
I'll try to be as simple as possible, explaining the steps but without going too much in details about "why" you need to do something (there's the Docker documentation for that).

I assume that you currently start the Docker container with a command like (with the addition of some flags and options)
```
docker run -it --rm irobotedu/create3-galactic bash
```

This command is saying: run a new container from the Docker image named `irobotedu/create3-galactic`.
If a Docker image with that name does not exist on your laptop, then Docker will look up for it on the Docker Hub and download it from there (this happened the first time you run the command).

In order to achieve your goal, you need to create a new Docker image that has some additions on top of `irobotedu/create3-galactic` (i.e. it changes the ROS 2 RMW).
Docker images are created using `Dockerfile` files.
A `Dockerfile` defines all the commands that need to be executed when creating the image (e.g. install ROS, install Fast-DDS, create directories, etc).

You can find here the `Dockerfile` of the Create 3 Docker image that you have been using so far https://github.com/iRobotEducation/create3-docker/blob/main/create3-galactic/Dockerfile

You can see that the first line says `FROM  osrf/ros:galactic-desktop`.
This indicates that the `irobotedu/create3-galactic` Docker image is created FROM the `osrf/ros:galactic-desktop` Docker image with the addition of some stuff on top of it.
This should sound familiar, as it's exactly what you are looking to do.

So you need to:
 1. Create a file named `Dockerfile` on your machine.
 2. Write the first line as `FROM  irobotedu/create3-galactic`
 3. Define your additional operations and commands in the Dockerfile.
 4. Build a Docker image from the Dockerfile using `docker build --tag my-create3-docker .` (note that the `.` at the end is not a typo, it defines the path of the Docker context; this command must be executed from the same directory where your new `Dockerfile` is located).
 5. Use your just built Docker image rather than the default Create 3 one: `docker run -it --rm my-create3-docker bash` (note how we use the same name passed to the `--tag` argument in the previous command).

This is the documentation of the syntax of Dockerfiles https://docs.docker.com/engine/reference/builder/

This is a Dockerfile that I think should be sufficient for your purposes, feel free to copy it and use it.
What it does should be pretty straight-forward.

```
FROM irobotedu/create3-galactic

RUN apt-get update && apt-get upgrade -y
RUN apt-get update && apt-get install -y ros-galactic-rmw-fastrtps-cpp

RUN echo ' \n\
export RMW_IMPLEMENTATION=rmw_fastrtps_cpp' >> $HOME/.bashrc
```

For completeness, this is not the only way to achieve your goal, but is the one I recommend.
A simpler, but hard to maintain, alternative would be to use the `commit` command https://docs.docker.com/engine/reference/commandline/commit/

**Comment by framlin** [comment]
Thank you very very much for your help!
